# Example usage of halo_trace 

First, to import the relevant packages

In [1]:
!ls

anaconda3			     halo_trace.h5  tangos_data.db
Anaconda3-2024.10-1-Linux-x86_64.sh  MAP	    tangos_halo_trace.h5
Desktop				     Music	    Templates
Documents			     Pictures	    Videos
Downloads			     Public


In [2]:
import os
os.chdir('/home/selvani/MAP/pynbody/halo_trace')
# !pip install -e .

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import halo_trace as ht
import pandas as pd 

%matplotlib inline

A bit about the tracing code:

The tracing code is fairly slow, but should be robust. By default it will cross check steps, meaning it calculates the main progenitor both 1 and 2 steps apart, to make sure they match. If it loses a main progenitor, it can also pick it up again in some cases. Because this process is so slow, usually what you want to do is run it once and save the output, then read in the output for all subsequent uses. Main progenitors are identified by maximizing a merit function (as in AHF's merger tree code), where the merit function of halo

\begin{equation}
\max_j\left(M_{ij} = \frac{N_{i \cap j}}{N_i N_j}\right),
\end{equation}

where $N_i$ and $N_j$ are the number of particles in halos in steps $i$ and $j$, N$_{i \cap j}$ is the number of shared particles. The merit function is maximized in both directions (step 1 to 2 and step 2 to 1), and a match is recorded only if the identified halos match.

Anyway...

The code traces halos and returns a pandas dataframe identifying the main progenitor at each step.

The only mandatory argument in the trace is `sim_base`, which is the path to the directory holding all your simulation time steps. You can specify a file name for the saved trace using `save_file`. You can trace specific halos by spcifying `grplist` (e.g., `grplist=[1, 3, 6]` will trace halos 1, 3, and 6 back), or you can trace all halos above some particle threshold. For all arguments, you can run `ht.tracing.trace_halos?`. So, if want to trace all halos with more than 1000 particles at z=0, no earlier than step 400, then this would run the trace (replacing `sim_base` with your own path):

## Trace

In [4]:
sim_base = '/home/selvani/MAP/Sims/cptmarvel.cosmo25cmb/cptmarvel.cosmo25cmb.4096g5HbwK1BH/snapshots_200crit_cptmarvel'
sim_base = '/home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/'
# ahf_dir = '.'
ahf_dir = 'ahf_200'

In [ ]:
# Test both simulation files AND halo catalogs comprehensively
# working_steps = ht.tracing.test_halo_catalogs_comprehensive(sim_base=sim_base, ahf_dir=ahf_dir)

Testing 62 snapshots for working simulation files AND halo catalogs...

--- Testing snapshot 1/62: 004096 ---
1. Loading simulation file...
   SUCCESS - Loaded <TipsySnap "/home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/h329.cosmo50PLK.3072gst5HbwK1BH.004096/h329.cosmo50PLK.3072gst5HbwK1BH.004096" len=55102465>
2. Testing particle data access...
   SUCCESS - 55102465 particles
3. Testing particle order array...
   SUCCESS - Order array accessible
4. Testing halo catalog loading...
   SUCCESS - Loaded 6670 halo groups

--- Testing snapshot 2/62: 004032 ---
1. Loading simulation file...
   SUCCESS - Loaded <TipsySnap "/home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/h329.cosmo50PLK.3072gst5HbwK1BH.004032/h329.cosmo50PLK.3072gst5HbwK1BH.004032" len=54950149>
2. Testing particle data access...
   SUCCESS - 54950149 particles
3. Testing particle order array...
   SUCCESS - Order array accessible
4. Testing halo catalog loading...
   S

pynbody.halo : Unable to load AHF substructure file; continuing without. To expose the underlying problem as an exception, pass ignore_missing_substructure=False to the AHFCatalogue constructor


   SUCCESS - Order array accessible
4. Testing halo catalog loading...
   SUCCESS - Loaded 733 halo groups

--- Testing snapshot 61/62: 000096 ---
1. Loading simulation file...
   SUCCESS - Loaded <TipsySnap "/home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/h329.cosmo50PLK.3072gst5HbwK1BH.000096/h329.cosmo50PLK.3072gst5HbwK1BH.000096" len=40809639>
2. Testing particle data access...
   SUCCESS - 40809639 particles
3. Testing particle order array...


pynbody.halo : Unable to load AHF substructure file; continuing without. To expose the underlying problem as an exception, pass ignore_missing_substructure=False to the AHFCatalogue constructor


   SUCCESS - Order array accessible
4. Testing halo catalog loading...
   SUCCESS - Loaded 597 halo groups

--- Testing snapshot 62/62: 000071 ---
1. Loading simulation file...
   SUCCESS - Loaded <TipsySnap "/home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/h329.cosmo50PLK.3072gst5HbwK1BH.000071/h329.cosmo50PLK.3072gst5HbwK1BH.000071" len=40809991>
2. Testing particle data access...
   SUCCESS - 40809991 particles
3. Testing particle order array...


pynbody.halo : Unable to load AHF substructure file; continuing without. To expose the underlying problem as an exception, pass ignore_missing_substructure=False to the AHFCatalogue constructor


   SUCCESS - Order array accessible
4. Testing halo catalog loading...
   SUCCESS - Loaded 477 halo groups

=== COMPREHENSIVE TEST SUMMARY ===
Total snapshots tested: 62
Working snapshots (both sim + halos): 62
Failed snapshots: 0

Working snapshots can be used with:
steplist = ['004096', '004032', '003936', '003840', '003744', '003648', '003606', '003552', '003456', '003360', '003264', '003195', '003168', '003072', '002976', '002880', '002784', '002688', '002592', '002554', '002496', '002400', '002304', '002208', '002112', '002088', '002016', '001920', '001824', '001740', '001728', '001632', '001536', '001475', '001440', '001344', '001269', '001248', '001152', '001106', '001056', '000974', '000960', '000864', '000776', '000768', '000672', '000637', '000576', '000480', '000456', '000384', '000347', '000288', '000275', '000225', '000192', '000188', '000139', '000107', '000096', '000071']


In [5]:
# working_steps = ht.tracing.test_halo_catalogs(sim_base=sim_base, ahf_dir=ahf_dir)


In [2]:
# sim_base = '/home/selvani/MAP/Sims/cptmarvel.cosmo25cmb/cptmarvel.cosmo25cmb.4096g5HbwK1BH/snapshots_200crit_cptmarvel'
sim_base = '/home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/'
# ahf_dir = None
ahf_dir = 'ahf_200'
# grplist = None
grplist = np.array([1])

In [3]:
trace = ht.tracing.trace_halos(sim_base=sim_base, 
                               ahf_dir=ahf_dir, 
                               grplist=grplist,
                               min_ntot=500)

62
Using only the following steps which contain ahf_dir and AHF catalogs:
['004096', '004032', '003936', '003840', '003744', '003648', '003606', '003552', '003456', '003360', '003264', '003195', '003168', '003072', '002976', '002880', '002784', '002688', '002592', '002554', '002496', '002400', '002304', '002208', '002112', '002088', '002016', '001920', '001824', '001740', '001728', '001632', '001536', '001475', '001440', '001344', '001269', '001248', '001152', '001106', '001056', '000974', '000960', '000864', '000776', '000768', '000672', '000637', '000576', '000480', '000456', '000384', '000347', '000288', '000275', '000225', '000192', '000188', '000139', '000107', '000096', '000071']
File will be saved as /home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/h329.cosmo50PLK.3072gst5HbwK1BH.004096/h329.cosmo50PLK.3072gst5HbwK1BH.004096.trace_back2.hdf5
grplist= [1]
simlow= /home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/h329.cosmo50P

Tracing halos:   0%|          | 0/61 [00:00<?, ?step/s]

Starting step 004032
simhigh= <TipsySnap "/home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/h329.cosmo50PLK.3072gst5HbwK1BH.004032/h329.cosmo50PLK.3072gst5HbwK1BH.004032" len=54950149>
Advanced step 003936
ahf_basename= /home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/h329.cosmo50PLK.3072gst5HbwK1BH.004032/ahf_200/h329.cosmo50PLK.3072gst5HbwK1BH.004032.0000.z0.015.AHF_
	Matching halos by merit function
1
[[13395302        0        0 ...        0        0        0]
 [       0   301186        0 ...        0        0        0]
 [       0        0   195401 ...        0        0        0]
 ...
 [       0        0        0 ...        0     7112        0]
 [       0        0        0 ...     7663        0        0]
 [       0        0        0 ...        0        0        0]]
	Ni= [13429482   301411   195401   163544   151125   132243   114018   116167
   110905    84241    95696    81763    61528    69911    58582    65514
    57577    5

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[13195165        0        0 ...        0        0        0]
 [       0   300360        0 ...        0        0        0]
 [       0        0   193245 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]]
	Ni= [13215356   300782   193245   159991   149427   131220   137682   113967
   109977    82871    94525    80251    60331    69064    56881    58708
    56928    35111    49478    50725    45433    41829    35859    36402
    36107    35039    35998    29001    32660    33171    28682    30028
    29060    22294    21543    24535    25049    25359    23900    19855
    20461    20972    19779    19627    18163    18182    17990    16131
    17992    17894    17040    16352    15637    15367    14856    14287
    14816    14743    14472    14003    14327    13743    13248    14032
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[13104895        0        0 ...        0        0        0]
 [       0   304084        0 ...        0        0        0]
 [       0        0   171151 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0     7836        0]
 [       0        0        0 ...        0        0     7743]]
	Ni= [13117594   304617   171151   158007   149168   140859   130125   113253
   109571    83363    94934    79482    59583    68539    55618    55354
    56735    48741    50456    44965    41704    36770    35324    24231
    35937    35850    34040    36856    35573    35808    35593    32475
    28555    23114    15128    24297    24837    25227    23703    19270
    20912    20013    20667    19504    17743    19344    18134    17860
    15770    18104    17062    16424    15443    15408    14358    14723
    14659    14685    14425    14274    14654    13830    13862    13299
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[13003073        0        0 ...        0        0        0]
 [       0   303202        0 ...        0        0        0]
 [       0        0   175395 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0     7827]
 [       0        0        0 ...        0        0        0]]
	Ni= [13014345   303838   185880   155965   147492   142694   128173   112914
   108670    83087    95362    78318    58600    68267    54344    56705
    53637    42079    48058    50333    44334    28411    40601    38086
    33745    35617    35220    37528    35537    33414    32226    27404
    28323    23687    24120    24570    25064    23574    18842    21449
    20742    19430    20335    15869    19460    15660    17806    17772
    18162    15523    16929    16525    15374    15305    14420    14587
    14590    15184    14654    14416    14453    14027    14657    13333
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[12923337        0        0 ...        0        0        0]
 [       0   298210        0 ...        0        0        0]
 [       0        0   240100 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0     7816        0]]
	Ni= [12950147   298849   240719   154074   146412   143960   126712   112287
   107806    82811    95573    76942    57729    67944    53325    56777
    52420    50120    47654    29153    43498    41506    39475    35019
    38821    32393    34709    35199    27631    33046    29913    32013
    24062    27890    23901    24350    24963    23386    21962    18362
    20611    20404    19417    19033        0    14278    17667    18130
    17354    15405    16896    16670    15249    15828    14495    15094
    15666    14537    14696    14576    14581    14311    14436    13359
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[12781809        0        0 ...        0        0        0]
 [       0   292055        0 ...        0        0        0]
 [       0        0   140790 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0     8031        0]
 [       0        0        0 ...        0        0        0]]
	Ni= [12821788   292916   157337   150577   144278   137861   124262   110643
   106998    80826    94067    75625    56732    67250    51955    56087
    51830    49627    46563    29259    41782    41269    38300    34119
    38821    31017    34219    34757    27085    32601    27578    31775
    23539    27518    23355    24061    24798    23110    21890    18007
    20436    20407    19088    18761        0    13468    17533    18029
    16977    15105    16603    16638    15343    15804    14325    14968
    15669    14277    14429    14348    14410    13340    14325    13203
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[12994152        0        0 ...        0        0        0]
 [       0   294570        0 ...        0        0        0]
 [       0        0   139896 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]]
	Ni= [13019611   295544   142519   154422   141032   146597   126093   112120
   107801    82956    98497    76156    57239    68341    53167    48995
    57721    51630    50200    47558    48226    43550    42362    42314
    40233    35178    35217    31187    35148    24151    35269    33136
    31900     1608    24834    27921    23663    24327    24960    23414
    23432    22163    20679    19357    19777    17921    18943    17816
    18174    15551    17472    16906    17021    16510    14930    14490
    15036    15311    15359    15409    15131    15031    14797    13449
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[12664722        0        0 ...        0        0        0]
 [       0   276555        0 ...        0        0        0]
 [       0        0   136094 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]]
	Ni= [12742030   280535   136518   145703   138567   141159   119166   108330
   105353    75554    94850    73019    54884    66559    50141    48078
    56154    50081    48756    45484    48226    41167    40164    41635
    37722    32936    33544    28187    34084    18678    35269    32097
    31545     1608    23217    27230    22909    23737    24519    22829
    23132    22038    20192    18763    19779    17442    18232    17174
    17924    14849    16601    16323    16923    16399    14513    13771
    14466    14744    14693    15578    14329    14579    14453    12972
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[12880611        0        0 ...        0        0        0]
 [       0   266859        0 ...        0        0        0]
 [       0        0   179957 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]]
	Ni= [12909105   270342   190156    34496   137162   146874   140772   118837
   108261   105266    99430    72336    71991    75268    67137    54552
    30462    50423    56752    49193    49632    46227    42769    42017
    42282    43698    38102    33738    33006    33954    32278    31790
    27710    23120    27374    26514    23252    23925    24603    23564
    22977    16819    21299    20165    19009    16918    17288    17951
    15147    16816    18067    14923    17531    16315    16674    14877
    14639    15127    14073    14232    14738    14752    14660    14253
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[12494002        0    44845 ...        0        0        0]
 [       0   247239        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]]
	Ni= [12602226   251731   137579    34496   134611   138744   135901   112615
   104520   102930    94678    68897    68945    74251    65282    52027
    42580    47566    54710    47819    48403    43827    40290    40607
    41427    43051    35841    32207    30774    32802    31295    31516
    25735    20543    26455    26288    22235    23381    24211    23162
    22391    16303    21140    19649    18448    15545    16753    17236
    14408    15950    17817    14295    17582    15626    16549    14279
    13823    14379    12661    13386    14309    14379    14147    13941
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[12838600        0        0 ...        0        0        0]
 [       0   249371        0 ...        0        0        0]
 [   78316        0        0 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...     8068        0        0]
 [       0        0        0 ...        0     7183        0]]
	Ni= [12854772   249672    78316   145670   132170   144989   138130   117505
    42689   106720   104523    97720    80342    66570    70393    66840
    53301    49437    39637    56151    48937    49663    42872    45075
    43875    44466    42216    37643    33501    32248    33683    32727
    32006    30731    27172    27248    21691    22885    23886    24499
    23463    22883    18887    21569    20628    17099    19910    14671
    14826    18719    18902    16935    17620    18070    16874    14761
    16949    14822    16133    14824    14615    14142    14644    14909
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[12724106        0        0 ...        0        0        0]
 [       0   245058        0 ...        0        0        0]
 [   75424        0        0 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0     7938]
 [       0        0        0 ...        0        0        0]]
	Ni= [12761897   245464    75424   140568   131034   142351   136475   116053
    42689   105747   103656    96450    79913    65959    69095    66292
    52685    48767    19036    55590    48595    49231    41932    44480
    43690    44251    41980    36800    32984    31683    33353    32648
    31712    30375    26695    26966    20882    22601    23724    24385
    23333    22717    18887    21503    20616    16748    19782    11726
    14974    18484    18902    16757    17442    17979    16566    14595
    16888    14587    15962    14572    14255    13972    14546    14811
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[12309955   377206        0 ...        0        0        0]
 [       0        0   239870 ...        0        0        0]
 [       0        0        0 ...        0        0     9280]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]]
	Ni= [12746725   240000    98826   129439   140036   133302    57131   117098
   104566   102990    92989    80673    65513    67883    65999    51875
    48558    54857    49069    48168    45606    43811    40432    44244
    41681    36064    33011    31303    32901    33724    31255    30490
    26723    26782    19487    24801    22309    20817    24301    23520
    23068    22497    21537    20267    19509    18213    18456    15653
    19677    16696    16825    16387    17946    17320     9418    14433
    16974    15117    14499    15608    14322    14426    14951    13137
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[11809267   459461        0 ...        0        0     3051]
 [       0        0        0 ...        0        0        0]
 [       0        0   148067 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]]
	Ni= [12356043   226598   151164   125930   132036   109468    57131   112654
   100912    99520    87168    78946    63373    64576    63937    49530
    45718    52879    47409    46568    45188    41048    34741    43298
    40803    33005    31524    29229    31627    33583    30091    29536
    24634    25850    17873    23991    21035    20787    23869    22845
    22580    21705    21316    20264    18883    18327    17567    14845
    19677    13782    15824    15654    17620    16813     3300    13803
    16728    14032    13674    15073    13444    14003    14512     9336
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[11520921    96634   111765 ...        0        0        0]
 [   17482   444852        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...     8817        0        0]
 [      39        0        0 ...        0        0        0]]
	Ni= [11737103   462334   148003   220049    90733   124762   127702   113442
    99438    96825    97650    79466    78730    62303    62648    62192
    48352    42628    51124    39606    46653    47538    45426    38937
    42844    40122    29211    32883    38556    31236    29144    30682
    29016    28635    23948    24001    24287    12990    23588    22055
    19055    18939    17322    22070    21133    21184    21742    20105
    19479    18230    17809    19115    16107    13699    10162    17410
    17451    16082    15270    16872    13413    14362    13019    14687
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[12110388    54610    30781 ...        0        0        0]
 [   61955   506185        0 ...        0        0        0]
 [    4296        0   197610 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0        0]]
	Ni= [12197720   568140   201906   229177   134328   114901   131063   115055
   109548    96163    95633    78156    79158    63194    66571    63115
    49386    41293    51425    47941    47966    39156    45342    43572
    39381    43350    40558    35376    32145    30192    30470    31620
    29904    28694    29227    28632    23920    24415    25159    23835
    23829    17811    23756    22268    21345    22091    21305    19075
    18195    18146    21240    18254    17150    17974    11280    16000
    17550    17402    13797    15469    11782    15287    14519    14175
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[11592018    36296    69853 ...        0        0        0]
 [  185750   380269        0 ...        0        0        0]
 [    4572        0   197291 ...        0        0        0]
 ...
 [       0        0        0 ...     8671        0        0]
 [       0        0        0 ...        0     8725        0]
 [       0        0        0 ...        0        0        0]]
	Ni= [11701474   566019   201863   218651    48029   109785   123181   107839
   104185    93833    91195    73481    75956    59815    64017    60483
    47468    38087    49243    47342    46116    35167    43713    41954
    36518    41876    39313    28286    29977    27484    30325    30257
    28355    26133    28087    27355    21652    23504    24161    22908
    23148        0    23367    21382    18841    21420    21001    18095
    16934    17095    21074    17729    16146    17476    12058    14633
    17173    17248    13144    14709     9860    14622    13890    13562
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[12199417    25969    18035 ...        0        0        0]
 [   73924   353522        0 ...        0        0        0]
 [    5614        0   262015 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0     8870]
 [       0        0        0 ...     8910        0        0]
 [       0        0        0 ...        0     7161        0]]
	Ni= [12244412   427446   267629   232941   128991   103115   103846   107012
    98598    92842    77803    77229    72537    62763    63360    50187
    50540    49386    39833    48055    40202    44977    37442    43301
    35641    40274    38904    30829    31538    34007    31471    29880
    29082    26091    29441    27321    27807    25654      843    23241
    25163    24403    24022    22302    21973    18866    21416    18901
    18208    13875    18346    17706    18726    18260    15668    17532
    17292    13975    15141    14863    14739    13831    13456    13052
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[11651351   336157    21069 ...        0        0        0]
 [     744   393296        0 ...        0        0        0]
 [   25081        0   254897 ...        0        0        0]
 ...
 [       0        0        0 ...        0        0        0]
 [       0        0        0 ...        0        0     6987]
 [       0        0        0 ...        0     8747        0]]
	Ni= [12017471   394040   279978   228140   125247   101419   105544    99103
    95799    90831    63755    76918    75748    73837    61670    62442
    49318    49625    48576    47824    38672    41150    44168    42812
    33928    39863    37984    36591    29810    30612    30855    29402
    28154    29110    27488    27338    25495    25115    25175    22487
    24363    23978    21962    21517    18257    21234    17986    18471
    12239    18049    18926    16665    18158    15217    17431    16867
    13683    14779    14470    14501    13500    13613    13300    13295
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[10705827   477374    16404 ...        0        0      438]
 [    1540   800974        0 ...        0        0        0]
 [   55263        0   218850 ...        0        0        0]
 ...
 [       0        0        0 ...        0     8427        0]
 [       0        0        0 ...     8556        0        0]
 [       0        0        0 ...        0        0        0]]
	Ni= [11204198   811470   274113   223829   119981    99611   103277    94204
    93545    87557    67456    73829    73005    59593    58038    63756
    61016    47528    47615    47870    47396    42316    37070    42833
    41846    39102    36172    29579    31477    28085    27293    29914
    14623    28498    29161    28496    26519    25214    24254    21438
    23760    24127    21546    21318    16528    20992    16759    16165
    19256    17697    17794    14657    15360    17246        0    16514
    13180    13493    14195    13679    14260    13127    12986    13460
   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[9872064  655580  209046 ...       0       0       0]
 [   2033  796394       0 ...       0       0       0]
 [   8465       0  267075 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [10744082   811014   275540   207199   112229    96953   101100    86868
    88863    83050    59934    70594    68604    56392    57974    63374
    58394    44906    43754    45645    46680    41485    34089    41097
    40269    38004    33179    30471    25233    19497    24623    27987
    22866    26928    28696    27355    25429    24315    22952    19708
    22991    23667    20724    19730    15253    20600    14018    14191
    18870    17078    17156    13782    14023    16799     6768    15253
    12367    12818    13489    12360    14260    12428    12398    13373
    12720    12044    11441     9078   

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[9593770   15603  126916 ...     242       0    1214]
 [   2271 1497277       0 ...       0       0       0]
 [    815       0  533852 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [9737957 1506358  534667  208054   98767  113776  103553   97187   93307
   86109   83451   72103   63900   60564   67722   57570   59317   44979
   43319   47019   45665   45040   35419   40894   40622   36043   38248
   36448   27756   27174   24809   27723   25539   25896   26197   23647
   21912   22806   23447   23636       0   19519   21069   20036   20820
   15402   14526   19910   16970   13551   17563   18141   16742   13794
   16960   15521   10643   12411   10371   12757   13477   14690   14309
   12813   12490   12216   12118   11661   12083   11664   11851   11757
   11337   10869   10929    9222   1063

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[9085381   25823  128199 ...     368       0       0]
 [   1491 1422417       0 ...       0       0       0]
 [   1393       0  531910 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [9242392 1439609  533303  194144  106607  106605  101583   94480   84729
   80260   76112   68908   55242   59327   63007   54068   55965   41704
   42109   46234   38316   42718   32726   38873   38770   26265   36866
   35996   25861   25581   23671   26441   21783   24687   24486   22072
   16477   21128   22616   23283   14399   17657   20209   18315   20301
   14300   14304   19565   16970   11979   16956   18052   15770   12238
   16565   14975    9644   11600       0   11278   12495   14768   14315
   12169   11650   11620   11441   11061   11619   10907   11368   11467
   10958   10401   10316    8475   1050

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[9355508       0     739 ...       0       0       9]
 [      0 1459917       0 ...       0       0       0]
 [    109       0  670065 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0    9143]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [9360964 1467472  670174  208547  109912  112979   98013   87622   84968
   80279   73392   68821   59474   61605   65713   57933   58848   45193
   42381   47596   43894   44537   36154   40291   40638   39769   37872
   32871   34222   27049   27183   27791   27792   26650   25049   24026
   23573   22428   24038   23399   19541   22651   21443   21649   18832
   16282   20932   16054   18473   15220   18372   18396   16170   17109
   13154   17105   16064   16430   12617   12402   11643   13226   12502
   13180   12498   12294   11408   12130   12036   11094   11988   12030
   11945   11585   10486   11564   1122

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[8974194       0    2527 ...       0       0       0]
 [      0 1391197       0 ...       0       0       0]
 [      0       0  655237 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [8992313 1407067  655237  196787  103816  106879   95715   80715   78960
   78208   69957   65850   43305   60033   62042   54801   56145   42783
   27176   46900   42836   42742   33282   38222   39154   38052   36725
   29570   30920   27048   25598   25752   26760   25730   23690   20233
   22276   21404   23410   22588   18715   21880   20627   20054   17485
   14008   20417   15206   16794   13149   17356   18271   15506   16500
   11235   16976   15355   16418   11270   11735   10646   12720   11870
   12400   12282   11958   10711   11682   11541   10339   11430   11778
   11506   11014    9617   11263   1066

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[8674088       0       0 ...       0       0       0]
 [      0 1331482       0 ...       0       0       0]
 [      0       0  647908 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [8679396 1343003  647908  192246   97036  102564   95033   74763   79690
   75574   67701   64457   58451   53360   59655   53873   41972   46823
   37773   42694   42377   41447   35176   31444   29801   31001   38492
   36645   34337   36094   23264   24541   26377   24397   26015   26259
   21945   23284   20884   21927   22118   17457   15373   20094   15707
   20056   20155   12470   16982   15103   18283   15174   17828   16379
   16841   15054   14556   12947   11345   10156    9985   13588   11703
   12703   10188   12009   11829   12314   12085   12101   10807   11377
   11587       0   11325   11239   1115

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[8159578       0       5 ...       0       0       0]
 [      0 1264666       0 ...       0       0       0]
 [      0       0  626799 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [8166421 1267450  626799  167966   86969   94837   91721   70703   70749
   69409   61384   59874   56589   49332   55215   50081   39023   45881
   31142   41375   42155   39169   34586   28248   29900   29902   36497
   34195   31805   34786   19755   22314   23144   21253   24692   24592
   15949   22515   19229   20797   21515   16908   12522   18881   14173
   18652   19394   10984   15788   14731   18165   13687   17454   15872
   16940   14776   13568   12111   10713    9631    8700   12962   11375
   11893    9474   11337   10619   11950    9915   11970    9797   10699
   11149       0   10880   10488   1065

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[8201661       0       0 ...       0       0       0]
 [      0 1235173       0 ...       0       0       0]
 [      0       0  636165 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [8201976 1265885  636165  162490   93121   96927   64803   69478   71678
   59551   61961   59583   57519   50136   57056   50592   38771   42724
   46603   44357   41889   41130   33309   31067   26136   29847   37037
   35462   34187   28039   20815   26582   27047   21942   23002   25250
   22209   18161   23551   23247   19919   18860   15849   20606   21349
   18913   19648   13447   18979   19027   14600   16002   15514   16044
   11402   14046   16070       0   11732   13672    7375   13365   12247
   11029   10302   12006    9109   12281   12531   12287   11539   11236
   12267   11237   10380   10172   1103

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[8136043       0       0 ...       0       0       0]
 [      0 1222512       0 ...       0       0       0]
 [      0       0  633137 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [8136393 1254722  633137  157979   92754   96041   71839   67649   71207
   50797   61474   59042   57267   49624   56468   50105   37999   42558
   46453   44269   40670   40818   32948   30088   23396   29419   36729
   35296   33878   25641   20386   26310   26748   21753   22717   24985
   22000   17862   23378   23161   19718   18867   15510   20435   21254
   18696   19566   13221   18880   19021   14544   15855   15172   15962
   11308   13923   16086   12161   11714   13542    7119   13213   12146
   10867   10158   11924    9046   12302   12419   12226   11420   11118
   12242   11150   10207   10078   1094

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[8005164       0       0 ...       0       0       0]
 [      0 1178964       0 ...       0       0       0]
 [      0       0  625326 ...       0       0       0]
 ...
 [      0       0       0 ...    6621       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [8006113 1186660  625326  159728   91446   94687   72226   55943   27940
   47332   62200   56581   57258   56345   47637   48746   42421   45918
   35245   44820   44407   40784   19939   27081   31740   35141   27722
   34774   32986   26195   18016   22126   19150   25876   24624   16222
   22517   23903   23419   23079   23214   17086   17349   13980   20930
   19492   19191   19847   18515   18342   18147   15122   13838   11090
   15730   13807   12293   10956   10681   14117   13209   14210   14380
   11925   13318    9891    7740   11598   12284   10245   11962   11243
   12191   10627   10910   10807   1136

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[7234317       0       0 ...       0       0       0]
 [      0 1090790       0 ...       0       0       0]
 [      0       0  597672 ...       0       0       0]
 ...
 [      0       0       0 ...    5946       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [7473593 1103055  597672  143911   87109   87281   66772   51612   32776
   45070   49895   53981   53055   51698   43016   45914   40567   44677
   30211   43927   38377   38175   10388   18251   31526   32460   24325
   33212   31024   22478   19734   21272   15378   22800   20989   15610
   19930   21757   21840   21923   22195   15295   17255    8769   20951
   18216   18192   19581   17482   17088   17049   14012   13215    9440
   14998   12928   11650    9605    7938   13329   11952   14067   14380
   10215   13812    9805    7940   10807   11859    9441   11397   10351
   11962    9980    9850   10112   1125

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[7640182       0       0 ...       0       0       0]
 [      0 1140215       0 ...       0       0       0]
 [      0       0  606363 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [7706813 1144587  606363  213157  122588   87468   91219   75763   30712
   59622   60407   55408   53798   43808   43537   47675   45533   20844
   45220   41454   45644   42648   40961   30046   28429   34362   33723
   26357   32727   25675   20587   22409   24186   24463   25524   21083
       0   22668   22586   22924   16738   21303   20388   15534   13516
   13957   13705   14213   18277   18587   18488   17843   17508   17949
   17469   15309   14086   14010   13014   14181   13649   12671   12155
    8113    7735   12429   12326   12265    8043   10786   11733   10748
    9147   10045   10395    9400    986

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[7391384       0       0 ...       0       0       0]
 [      0 1103799       0 ...       0       0       0]
 [      0       0  593986 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [7482491 1107963  595931  193369  110860   85538   87029   73003   31446
   56852   57665   54102   51798   41934   43497   45634   43138   20844
   44645   40485   45048   40400   39502   29014   22142   33631   33104
   25137   31522   25256   16888   20656   23008   23606   25525   20066
       0   21988   21844   22331   15471   19939   19398   11208   13252
   13114   11578   13693   17404   17988   18052   17782   16891   17507
   17472   15014   12945   13568   11427   13862   13340   12140   11705
    6965    6837   12268   12160   12181    7575   10317   11482   10466
    8554   10067   10145       0    950

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[6167782       0  874296 ...       0       0       0]
 [      0 1031514       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [7113742 1039974  575523  107178   82170   81621   66583   40570   53269
   57225   52545   42402   42858   44804   49439   44070   39358   39632
   44035   39241   35637   41690   35500   34556   28228   27676   32699
    3275   29429   23059   25846   23710   18352   16397   21785   23186
   21561   18134   21129   21449   20734   23546   14199   17645   16828
   17330   17220   16679   16083   17177    9118   14792   13922   10970
   12567   11457   13392   12817   13397   11144    7532   11073    9726
    9277   13010   12189   12542   11268    9518    7539    9982    8939
    9664    9908    9279    8732    978

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[5589028       0  853729 ...       0       0       0]
 [      0  935302       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0    7285]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [6649479  962166  544904   86213   78887   73248   54166   40570   47340
   52208   49952   42350   42562   42282   45618   41423   35920   33735
   42520   37437   38167   41233   27286   29922   23230   24986   30785
   16811   27239   20015   26420   22951   17350   10024   19322   22059
   19882   16307   19512   20133   18339   22524   11784   16175   15079
   16191   16360   15721   14859   16231    8697   14338   13834    9343
   11691   10891   12502   12123   12634    9879   10374   10337    7781
    8846   12982   11746   12460   10691    7518       0    9406    7849
    8970    9142    8456    8234    906

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


ahf_basename= /home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/h329.cosmo50PLK.3072gst5HbwK1BH.001248/ahf_200/h329.cosmo50PLK.3072gst5HbwK1BH.001248.0000.z1.529.AHF_
	Matching halos by merit function
1
[[5819969       0    2098 ...       0       0       0]
 [      0  979704       0 ...       0       0       0]
 [     10       0  794096 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0    8383]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [5847449  980722  866530  558655  112023   81749   81466   84107   71232
   62331   49743   55288   59008   32854   52199   52023   42563   43255
   43171   39394   42266   38565   39622   38484   34824   31967   27962
   32237    3417   28584   27726   25068   25566   24914   22211   24704
   22813   22724   20216   22044   22254   16384   21231   21644   17450
   20001   14075   18717   17019   17638   17035   17262   17079

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[5200005       0       0 ...       0       0       0]
 [      0  883548       0 ...       0       0       0]
 [      1       0       0 ...       0       0       0]
 ...
 [      0       0       0 ...    7486       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [5296143  885950  729305  513849  111916   84290   76953   72730   67001
   49647   49114   42818   54869   32831   48341   45936   40005   37804
   40869   36737   36030   34177   35145   25533   29648   22377   25416
   29454   10539   25576   25937   23515   22183   24583   19841   21675
   16438   22419   20155   17964   19507   13527   15975   18187   17450
   17894   11180   17590   15564   16173   15035   15552   15862   15689
   14425   12786   14682    8323   14474   13868   15018   13080    7879
    7843   11595   11852   12022   11230    7434    8983   10063    9575
    8762    9131    8757    6232    814

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[5467345       0       0 ...       0       0       0]
 [      0  900133       0 ...       0       0       0]
 [      0       0  506847 ...       0       0       0]
 ...
 [      0       0       0 ...    7495       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0    7152       0]]
	Ni= [5502809  900636  506852  419491  268779  145573   77829   78367   77840
   66860   56081   47388   55884   56207   36875   48606   38025   46102
   38437   40675   37678   41265   35810   33843   35405   28426   28064
   29729   28199   29665   25005   14150   23611   26283     997   24013
   20569   20790   22087   22026   22587   20042   19477   17649   18153
   18557   15637   12984   15631   16912   16080   16533   16281   16455
   16110   15348   15679   15790   10017   15319   14086   12525   13396
   12423   12412   12031    9324   11231    9873    8491    9323    7836
       0    8670    6827    9958    904

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[5136799       0       0 ...       0       0       0]
 [      0  857182       0 ...       0       0       0]
 [      0       0  479932 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0    6988]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [5197570  858967  479996  399305  264270  144941   77727   75801   71494
   63592   56066   38485   52419   51970   28321   46157   31624   41541
   39435   39713   36634   39384   33269   27353   34263   26094   25118
   28066   25187   28496   23957   12562   23623   24245     863   23707
   18399   20784   19115   20231   20914   18412   18266   16494   17360
   18511   14083   11771   15274   16216   15248   15744   15239   15865
   15852   14626   14741   15083   10017   14499   13237   11573   12422
   11576   11737   11288    8079    9978    9380    6510    8002       0
       0    7530    5441    9397    828

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[4989910       0       0 ...       0       0       0]
 [      0  832268       0 ...       0       0       0]
 [      0       0  459039 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [5134399  835409  459151  368756  285704  170904  102096   74471   61653
   60883   57515   50957   49157   44483   37927   40462   39170   15507
   32199   39116   23311   32319   33816   24860   27572   30819   21054
   28583   21749   23631   22913   23298   23056   21494   22055   17620
   20123   16782   17937   21266   13507   17042   17208   16794   17027
   14831   14193   14388   10358   15017   13024   13523   14984   14743
   12744   14216   12867   13392   11299   12503    9711   11229   12941
   11155   11306   10998   10595    9918   10933   10075   10485    9126
    9688    7415    8959    7153    742

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[4856519       0       0 ...       0       0       0]
 [      0  812644       0 ...       0       0       0]
 [      0       0  452889 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [5032392  815866  453035  363630  280020  170948  101802   73337   60343
   60101   57515   49745   48256   43500   37761   39983   39025   20226
   30981   38191   23287   31412   33094   24544   25904   30808   20478
   28342   21600   23234   22913   23102   22486   20797   21634   17500
   19760   16267   18475   21121   12484   16680   16940   16527   16771
   14428   13782   13882    9854   14723   12533   12774   14835   14460
   12480   13965   12489   13124   10865   12257    9826   10699   12913
   10854   10949   10776   10343    9760   10905    9962   10445    8927
    9637    7079    8840    7034    724

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[4255634       0       0 ...       0       0    1566]
 [      0  738811       0 ...       0       0       0]
 [      0       0  448528 ...       0       0       0]
 ...
 [      0       0       0 ...    6399       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [4434220  739064  449010  345202  250142  193468  185365  102362   68121
   54441   39658   48600   35499   46810   41077   46021   38555   35423
   28143   33113   30872   26962   16896   29026   21070   17517   21747
   22818   17175   22589   10955   21132    9248   19716   20948   19127
   16994   14539   18064   10949   15384   15095   15443   15684   14487
   13699   13866   12207    7494   11539   13764   13082   13404   10353
   10900    8539   11601    8639   12683   10512   11305    7281    9355
   11517   10423   10663   11540    9583   10530    9883    9246    7842
    7893    9305    8893    7659    614

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[3662740       0       0 ...       0       0       0]
 [      0  631037       0 ...       0       0       0]
 [      0       0  418894 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0    5380]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [3918798  631473  419807  319897  224371  208644  176876   97952   62059
   47537   36378   42849   26109   38275   36644   41881   34007   29182
   25533   31802   24824   26431   13569   20014   20488   12519   20059
   19714   16366   21525    5741   13325    5236   15590   17514   16009
   14410   12543   15529    9730   12756   12855   13282   13928   12633
   11138   11266   10660    6886    9053   12454   11123   12439    8820
    8140    7110   10128       0   11309    9109   10639       0    8455
   10469    8676    8834   11575    8195   10530    8931    7918    6494
    7272    8041    7567    6133    403

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


ahf_basename= /home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/h329.cosmo50PLK.3072gst5HbwK1BH.000768/ahf_200/h329.cosmo50PLK.3072gst5HbwK1BH.000768.0000.z2.524.AHF_
	Matching halos by merit function
1
[[4052953       0       0 ...       0       0       0]
 [      0  731914       0 ...       0       0       0]
 [      0       0  531642 ...       0       0       0]
 ...
 [      0       0       0 ...       0    5625       0]
 [      0       0       0 ...    6602       0       0]
 [      0       0       0 ...       0       0    6360]]
	Ni= [4056922  731914  531814  359875  351531  258507  220015  122827   65769
   66716   42118   41145   52713   49789   40584   37079   34235   41469
   30671   34965   28187   26365   19040   20572   21869   24101   24102
   20689   23282   21985   17709   17200   15312   13705   18164   17097
   15891   16017   14857   15523   14911   13504   14151   13583   13253
   14378   13341   14283   14142   11039   13581   12530   12497

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[3099005       0       0 ...       0       0       0]
 [      0  622534       0 ...       0       0       0]
 [      0       0  433031 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0    5473       0]]
	Ni= [3291215  622534  441039  290923  302465  220729  182978  108404   57674
   61428   33740   33747   44392   41039   29384   31340   30341   26772
   24268   21658   22083   20076   16923   16699   18611   22564   17815
   16588   17788   16910   15488   17440   12431   10238   14523   14465
   11773   12363   14641   13684    9826   11432   12675    7827   10884
   11242       0    8873    9142   10460   11110   10192   11481   10600
    6784    9137   10477    8559    8489    7758    7597    9040    7834
    9241    7997    6731    7412    7843    7894    7625       0    4562
    6924    7854    6640    6551    658

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[2827385       0       0 ...       0       0       0]
 [      0  746430       0 ...       0       0       0]
 [      0       0  427575 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [2882070  746430  427928  306038  280888  232839  170468   98102   82026
   62809   69161   58605   49073   29816   33868   42067   20953   29846
   32274   19257   26562   26399   18460   22978   28582   20257   23832
   10656   23457   14107   20503   18941   16851   13977   11954   14272
   14548   10818   15103   12860   14932   11223   13393   13045   11076
   12020   11713   10111    9574   11721   10776   11197    7333   11109
   10463   10346    9014    8488    8798    9882    9527    8962    9255
    8220    5843    8088    7431    7993    7271    8056    6706    8037
    7957    8048    8208    7646    705

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[1863495       0  509490 ...       0       0       0]
 [      0  602710       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [2444932  602710  294455  268378  224292  195705  155687   95000   76917
   54139   60241   54003   42729   23236   29884   36114   12585   23652
   28679   15142   24330   25052   16567   19270   24846   17356   21780
   15800   18644   11088   11278   14696   14281   11773    8620   12273
   10874    7913   12287   10568   13086   10274   11785   11361    9401
    9378    9657       0    7911   10440    8186    9861       0    9844
    8756    7815    6856    8700    8737      20    8264    7112    8023
    6624    3805    5976       0    6416       0    7051       0    6650
    7079    6883    7268    6702    612

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


ahf_basename= /home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/h329.cosmo50PLK.3072gst5HbwK1BH.000480/ahf_200/h329.cosmo50PLK.3072gst5HbwK1BH.000480.0000.z3.831.AHF_
	Matching halos by merit function
1
[[1083347       0    1309 ...       0       0       0]
 [      0  390764       0 ...       0       0       0]
 [   2919       0  227875 ...       0       0       0]
 ...
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]
 [      0       0       0 ...       0       0       0]]
	Ni= [1389308  530481  338031  215503  154831  150757  124463  158410   77251
   68891   50391   48065   48455   51234   33384   26037   20776   32748
   26698   24233   22639   24142   19937   18978   19886   13669   15387
       0   15164   10220   13082    8475    9324   10690    9309   11794
    9869   10972   11592   11532   11840    9639    9081    9233    5630
    9424       0    7231    8228    8149       0    5309    8596

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[966190      0   2279 ...      0   4732      0]
 [     0 372693      0 ...      0      0      0]
 [  3282      0 221201 ...      0      0      0]
 ...
 [     0      0      0 ...      0      0      0]
 [     0      0      0 ...      0      0      0]
 [     0      0      0 ...      0      0      0]]
	Ni= [1276666  517014  347808  204798  151168  138921  119118  152256   75664
   64433   49035   46435   46282   48688   29170   21840   18562   30086
   26114   22266   21726   22611   18835   18063   17649   12564   14526
       0   14155    9320   12173    7547    7277    8069    7406   10841
    9150   10645   11282   11095   11498    9313    8522    8763    4899
    9283       0    6984    7574    7639       0    4880    8183    7296
    8599    6659       0    5617       0    8107    6118    5170    5806
    5295    6100    6506       0    8642    3808       0    5488    5156
    5975       0       0    6114       0       0    5471    5155    6212
  

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[651627      0      0 ...      0      0      0]
 [     0 281364      0 ...      0      0      0]
 [     0      0 184646 ...      0      0      0]
 ...
 [     0      0      0 ...      0      0      0]
 [  4538      0      0 ...      0      0      0]
 [     0      0      0 ...      0      0      0]]
	Ni= [713043 356426 185172 160277 112465 137890 129118 125147 101295  77865
 102878  74077  54851  50802  47489  31132  42212  35339  36928  20894
  25459  21910  21861  17198  18524  19052   9048  16570  15185  13764
  13092  14085  11104  10390  12592  12885  11695   8372  10115  10346
   8476   8620   6344   9275   8689   9537   9349   8847      0   8796
   6840   6686   3959   7543   7645   4355      0   7256   6247   6579
   5585   4616   6208   5938   6017   4909   5266   5554   5312   5842
   6106   5619   5426   6368   5250   5260      0      0   5094      0
   5252   4784      0   5987      0   4987   4581   5583   5348   4998
      0   4869      

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[574907      0      0 ...      0      0      0]
 [     0 231924      0 ...      0      0      0]
 [     0      0 166617 ...      0      0      0]
 ...
 [     0      0      0 ...      0      0      0]
 [    75      0      0 ...      0      0      0]
 [     0      0      0 ...      0      0      0]]
	Ni= [640183 325006 167216 135080 109955 119527 122972 103960  97072  63744
  85103  65459  50967  47407  43370  25643  36118  28070  34226  18782
  22686  19989  18401  15500  15586  17579   7716  15695  14071  10819
  11524  13996   9956   8773  11828  11408   6095   6112   5474   8619
   7031   7809   4963   8661   7598      0   8267   6409      0   7449
   6226   5667   4932   6955      0   3063      0   6918   5497   5875
   5159      0   5370   5392   5300      0   4710   4964   4705   5236
   5637   5116   4336   6173      0   4353      0      0   4606      0
   3660   4401      0   5348      0   4707      0   5583   4767   4250
      0   4286      

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[451357      0      0 ...      0      0      0]
 [     0 126852      0 ...      0      0      0]
 [     0      0      0 ...      0      0      0]
 ...
 [     0      0      0 ...      0      0      0]
 [     0      0      0 ...      0      0      0]
 [     0      0      0 ...      0      0      0]]
	Ni= [460777 131511 126797 114020      0  88778  86086  45033  55612  33067
  52784  45665  38694  43184  19806  28596  18084  26471  29746      0
  20734  18290  19868   3398  18649  17598  13159  17781  17845  16594
  13983  13838  13251  12752      0  11129   9189  12389  11490  10611
  11964   9647   9342   9021   7871   7008   5292   7729   5112   6824
   7641      0   5843      0   4604   4307   3548   5414   5440   6252
   4649   5148   4680      0   4853   2970   4432   4893   4355   5431
   5366   4697   4954   4101   4874   4115      0   3858      0      0
   4336   5013   3315   3489   3881   2593      0      1   3852   3967
   3752   3405      

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[416925      0      0 ...      0      0      0]
 [     0 122678      0 ...      0      0      0]
 [     0      0 113469 ...      0      0      0]
 ...
 [     0      0      0 ...      0      0      0]
 [     0      0      0 ...      0      0      0]
 [     0      0      0 ...      0      0      0]]
	Ni= [431421 136288 117102 100319  74921  81603  79838  31104  50337    512
  48861  42904  32817  43277  17806  21167  13619  24248  27372      0
  17846  16051  18198   3305  17987  16465  12723  16328  16520  15995
  13286  12999  12065  12198   9208  10324   9087   9578   9686  10283
  11377   9017   8758   8677   7402   6291   4485   7430   4435   5938
   6581      0   5500      0   3978      0   3279   4646   4989   6387
   4261   5113      0      0   4549   2507   3797   3709   3923   5087
   5142   4466   3867   3635   4682   3789      0   3630      0      0
   3957   4781      0   3725   3733   2487      0      1   3743   3889
   3423      0      

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[217073      0      0 ...      0      0      0]
 [     0      0      0 ...      0      0      0]
 [     0  89311      0 ...      0      0      0]
 ...
 [     0      0      0 ...      0      0      0]
 [     0      0      0 ...      0      0      0]
 [     0      0      0 ...      0      0      0]]
	Ni= [267619  74178  89361  51632  55832  13441  35847  45844  32401  27430
  21089  28031  21365  24768  23882  12504  17126  15188  15317  12873
  12747   9128   9186  11853  14480  11583   6641   5814      0  10537
   9949      0   8198   8841   8510   7847   8310   7337   6613   7869
   5693   3774   6746   3192      0   6258   3006      0   5480   4328
      0      0   4908      0   4799      0   3674      0   3504   2936
   3726   3439   3330      0   3660   3126   2231      0      0   3213
   3234   1982   3771   3320      0      0   2958   3170   2748      0
   3043      0      0   2419      0      0   2713      0      0      0
      0      0   276

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
1
[[    0     0     0 ...     0     0     0]
 [    0     0     0 ...     0     0     0]
 [75434     0     0 ...     0     0     0]
 ...
 [    0     0     0 ...     0     0     0]
 [    0     0     0 ...     0     0     0]
 [    0     0     0 ...     0     0     0]]
	Ni= [67313 44998 75504 27845 42966 20408 33392 32855 28552 20718 19690    25
 15967 22162 24370  9084 13146 13052 15238 10036 10817  1703  7944 10212
 13891  9845  5945  5291  5671 10523  7084     0  5312  4994  2785  7749
     0  5621  5737  7730  3606     0  4779  2131     0  3953     0  2588
  3193     0     0     0  4011     0  4273     0  3063     3  2738     0
  3353  2215  3015     0  3546     0  2690     0     0     0  2372     0
  3142  2674     0  2205  2168  3409  2586  1863     0     0     0     0
     0     0  2765     0     0     0     0     0  2798     0     0     0
     0     0  2020     0]
	Nj= [75434 35424 24370 33392 28552 26136 24854 22162 19715 22019 20683 18573
 15238 

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


ahf_basename= /home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/h329.cosmo50PLK.3072gst5HbwK1BH.000188/ahf_200/h329.cosmo50PLK.3072gst5HbwK1BH.000188.0000.z7.986.AHF_
	Matching halos by merit function
36
[[85982     0     0 ...     0     0     0]
 [    0     0     0 ...     0     0     0]
 [    0     0 44210 ...     0     0     0]
 ...
 [    0     0     0 ...     0     0     0]
 [    0     0     0 ...     0     0     0]
 [    0     0     0 ...     0     0     0]]
	Ni= [85982 32084 44210 37484 32574 28140 23395 26737 23041 20269 13555 18069
 17861 16131 13525 13059     0 10685 11197 11513 11194 10455 10427  9964
 10389  9436  9428  8800  8640   341     0     0  7178  6871  6347  6379
  5994  5682  5612  5797  5439  5256  4679  4626  4740  4389  4272  4008
  4174  4328     0  3375  2771  4059  3813  3379  3513  3532  3576  3678
  2912  3537  3186  2812  3313  3096  3192  3141  2946  2761  2343  2672
  3054  2629  2545  2495  2566  2447  2456     0  2508  2333  

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)


	Matching halos by merit function
9
[[43980     0     0 ...     0     0     0]
 [    0 26172     0 ...     0     0     0]
 [    0     0     0 ...     0     0     0]
 ...
 [    0     0     0 ...     0     0     0]
 [    0     0     0 ...     0     0     0]
 [    0     0     0 ...     0     0     0]]
	Ni= [43980 26359 20659 17702 17944 14386  9790 14530     0  4201  4598  7903
 11062 10477  7352     0  7084  3132  5780  6386  5268     0  9183  3751
  4356  3814  3347  6466  2320  6364  3637  4181  4204  1528     0  3775
  2713     0  2612     0  2195     0     0  1652     0  2837  1577  1650
     0  1939     0  2080  1750  2719     0  1801  2295  2092     0     0
     0  2282  1720     0     0     0     0     0  2419     0  1652  1174
  1714  1145     0  1526     0     0  1378   986     0     0  1194     0
     0     0  1591     0     0  1270  1714     0  1095     0     0     0
     0     0  1174     0]
	Nj= [43980 26347 17944 20659     0     0 14530  9183 12769 12825 11062  6466
  6376 

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)
pynbody.halo : Unable to load AHF substructure file; continuing without. To expose the underlying problem as an exception, pass ignore_missing_substructure=False to the AHFCatalogue constructor


	Matching halos by merit function
46


/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)
pynbody.halo : Unable to load AHF substructure file; continuing without. To expose the underlying problem as an exception, pass ignore_missing_substructure=False to the AHFCatalogue constructor


[[   0    0    0 ...    0    0    0]
 [   0    0    0 ...    0    0    0]
 [   0 9049    0 ...    0    0    0]
 ...
 [   0    0    0 ...    0    0    0]
 [   0    0    0 ...    0    0    0]
 [   0    0    0 ...    0    0    0]]
	Ni= [    0     0  9049     0     0     0 10342     0     0     0     0     0
     0     0     0  3399     0     0     0     0     0     0     0     0
  1697  2360     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0  1452     0     0
     0     0     0     0     0     0     0     0   551     0     0     0
     0     0   799     0     0     0     0     0     0     0   448   522
   454     0   234     0     0     0     0     0     0     0     0     0
     0     0     0     0]
	Nj= [10342  9049  3399  2360  1697  1452     0     0     0   522   551   610
     0     0     0     0     0     0   454     0     0     0     0     0
     0

pynbody.halo : Unable to load AHF substructure file; continuing without. To expose the underlying problem as an exception, pass ignore_missing_substructure=False to the AHFCatalogue constructor


[[   0    0    0 ...    0    0    0]
 [   0    0    0 ...    0    0    0]
 [   0 2836    0 ...    0    0    0]
 ...
 [   0    0    0 ...    0    0    0]
 [   0    0    0 ...    0    0    0]
 [   0    0    0 ...    0    0    0]]
	Ni= [   0    0 2836    0    0    0 8370    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0  447    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0 1247    0    0    0    0    0    0    0
    0 1443    0    0    0    0    0    0    0    0    0    0  769    0
    0    0    0    0  482    0    0    0    0    0    0    0    0  558
    0    0    0    0    0    0    0    0    0    0  576    0    0    0
    0    0]
	Nj= [8370 2836 1443 1247  769    0  558    0  576    0    0  447    0    0
  482    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)
pynbody.halo : Unable to load AHF substructure file; continuing without. To expose the underlying problem as an exception, pass ignore_missing_substructure=False to the AHFCatalogue constructor


[[8311    0    0 ...    0    0    0]
 [   0 2697    0 ...    0    0    0]
 [   0    0    0 ...    0    0    0]
 ...
 [   0    0    0 ...    0    0    0]
 [   0    0    0 ...    0    0    0]
 [   0    0    0 ...    0    0    0]]
	Ni= [8311 2697    0  447    0 1520  530  496  479  430  473  450  462  411
  365  325  311  391    0  347  360  258  340    0  327    0  322    0
  327  228  238  286  275  273  297  279  299  261  280  234  256  243
  198  237    0  248  270  236  211  203  243  223  256    0  232  252
  197  227  227  191    0  214    0  200  188  204  229  207    0    0
    0  203    0    0    0  186    0  187  191    0    0    0  183  187
    0    0    0  175    0  185  186    0    0    0  167    0  198  178
  181    0]
	Nj= [8311 2697 1520    0  473  530  430  496    0  479  462  447  411    0
  450  360  322  340  391  347  327  365  275  311  279  270  327    0
  286  325  297  280  252  299    0  256  261  273  229    0  214  256
    0  248  227    0  227  236  243  232

/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:81: RuntimeWarning: invalid value encountered in divide
  merit_f = mat**2/np.outer(Ni, Nj)
/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:278: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '[nan]' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.loc[mask, step[-6:]] = df2.loc[mask, step[-6:]]
/home/selvani/MAP/pynbody/halo_trace/halo_trace/tracing.py:293: PerformanceWarning: 
your performance may suffer as PyTables will pickle object types that it cannot
map directly to c-types [inferred_type->floating,key->block1_values] [items->Index(['000071'], dtype='object')]

  df.to_hdf(save_file, key='ids')


In [5]:
'''# Debug: Check what paths are being constructed
from pathlib import Path
import glob

sim_high_path = '/home/selvani/MAP/Sims/cptmarvel.cosmo25cmb/cptmarvel.cosmo25cmb.4096g5HbwK1BH/cptmarvel.cosmo25cmb.4096g5HbwK1BH.003968/cptmarvel.cosmo25cmb.4096g5HbwK1BH.003968'
pth = Path(sim_high_path)
ahf_dir = 'ahf_200'

print(f"Simulation file path: {pth}")
print(f"Parent directory: {pth.parent}")
print(f"Looking for AHF files in: {pth.parent / ahf_dir}")

# Check what files actually exist
ahf_files = list(pth.parent.glob(f'{ahf_dir}/*AHF_halos'))
print(f"Found AHF files: {ahf_files}")

if ahf_files:
    ahf_basename = str(ahf_files[0])[:-10]  # Remove '.AHF_halos'
    print(f"ahf_basename would be: {ahf_basename}")
    
    # Check what related files exist
    print(f"Checking for related files:")
    print(f"  {ahf_basename}.AHF_halos exists: {Path(ahf_basename + '.AHF_halos').exists()}")
    print(f"  {ahf_basename}.AHF_particles exists: {Path(ahf_basename + '.AHF_particles').exists()}")
'''

'# Debug: Check what paths are being constructed\nfrom pathlib import Path\nimport glob\n\nsim_high_path = \'/home/selvani/MAP/Sims/cptmarvel.cosmo25cmb/cptmarvel.cosmo25cmb.4096g5HbwK1BH/cptmarvel.cosmo25cmb.4096g5HbwK1BH.003968/cptmarvel.cosmo25cmb.4096g5HbwK1BH.003968\'\npth = Path(sim_high_path)\nahf_dir = \'ahf_200\'\n\nprint(f"Simulation file path: {pth}")\nprint(f"Parent directory: {pth.parent}")\nprint(f"Looking for AHF files in: {pth.parent / ahf_dir}")\n\n# Check what files actually exist\nahf_files = list(pth.parent.glob(f\'{ahf_dir}/*AHF_halos\'))\nprint(f"Found AHF files: {ahf_files}")\n\nif ahf_files:\n    ahf_basename = str(ahf_files[0])[:-10]  # Remove \'.AHF_halos\'\n    print(f"ahf_basename would be: {ahf_basename}")\n\n    # Check what related files exist\n    print(f"Checking for related files:")\n    print(f"  {ahf_basename}.AHF_halos exists: {Path(ahf_basename + \'.AHF_halos\').exists()}")\n    print(f"  {ahf_basename}.AHF_particles exists: {Path(ahf_basename + \'

In [6]:
'''import pynbody as pb
s = pb.load(sim_base + 'cptmarvel.cosmo25cmb.4096g5HbwK1BH.004096/cptmarvel.cosmo25cmb.4096g5HbwK1BH.004096')
s.physical_units()

ahf_basename = sim_base + '/cptmarvel.cosmo25cmb.4096g5HbwK1BH.004096/ahf_200/cptmarvel.cosmo25cmb.4096g5HbwK1BH.004096.0000.z0.000.AHF_'
#check if file exists
if not os.path.exists(ahf_basename + 'halos'):
    raise FileNotFoundError(f"AHF halos file does not exist: {ahf_basename}.AHF_halos")
else:
    print(f"AHF halos file found: {ahf_basename}halos")
halos = s.halos(halo_numbers='v1', ahf_basename=ahf_basename)
print(f"Loaded halos: {halos}")'''

'import pynbody as pb\ns = pb.load(sim_base + \'cptmarvel.cosmo25cmb.4096g5HbwK1BH.004096/cptmarvel.cosmo25cmb.4096g5HbwK1BH.004096\')\ns.physical_units()\n\nahf_basename = sim_base + \'/cptmarvel.cosmo25cmb.4096g5HbwK1BH.004096/ahf_200/cptmarvel.cosmo25cmb.4096g5HbwK1BH.004096.0000.z0.000.AHF_\'\n#check if file exists\nif not os.path.exists(ahf_basename + \'halos\'):\n    raise FileNotFoundError(f"AHF halos file does not exist: {ahf_basename}.AHF_halos")\nelse:\n    print(f"AHF halos file found: {ahf_basename}halos")\nhalos = s.halos(halo_numbers=\'v1\', ahf_basename=ahf_basename)\nprint(f"Loaded halos: {halos}")'

In [7]:
'''import pynbody as pb
file = sim_base + '/snapshots_200crit_cptmarvel/cptmarvel.cosmo25cmb.4096g5HbwK1BH.003968/'
s = pb.load(file)
h = s.halos(halo_numbers='v1')
# Check if the halo catalog is loaded correctly
if h is None:
    raise ValueError("Halo catalog could not be loaded. Please check the file path and format.")
# Check the number of halos
num_halos = len(h)
print(f"Number of halos in the catalog: {num_halos}")'''

'import pynbody as pb\nfile = sim_base + \'/snapshots_200crit_cptmarvel/cptmarvel.cosmo25cmb.4096g5HbwK1BH.003968/\'\ns = pb.load(file)\nh = s.halos(halo_numbers=\'v1\')\n# Check if the halo catalog is loaded correctly\nif h is None:\n    raise ValueError("Halo catalog could not be loaded. Please check the file path and format.")\n# Check the number of halos\nnum_halos = len(h)\nprint(f"Number of halos in the catalog: {num_halos}")'

In [8]:
'''import pynbody as pb
s = pb.load(sim_base + 'cptmarvel.cosmo25cmb.4096g5HbwK1BH.003968/cptmarvel.cosmo25cmb.4096g5HbwK1BH.003968')
s.physical_units()

ahf_basename = sim_base + '/cptmarvel.cosmo25cmb.4096g5HbwK1BH.003968/ahf_200/cptmarvel.cosmo25cmb.4096g5HbwK1BH.003968.0000.z0.033.AHF_'
#check if file exists
if not os.path.exists(ahf_basename + 'halos'):
    raise FileNotFoundError(f"AHF halos file does not exist: {ahf_basename}.AHF_halos")
else:
    print(f"AHF halos file found: {ahf_basename}halos")
halos = s.halos(halo_numbers='v1', ahf_basename=ahf_basename)'''

'import pynbody as pb\ns = pb.load(sim_base + \'cptmarvel.cosmo25cmb.4096g5HbwK1BH.003968/cptmarvel.cosmo25cmb.4096g5HbwK1BH.003968\')\ns.physical_units()\n\nahf_basename = sim_base + \'/cptmarvel.cosmo25cmb.4096g5HbwK1BH.003968/ahf_200/cptmarvel.cosmo25cmb.4096g5HbwK1BH.003968.0000.z0.033.AHF_\'\n#check if file exists\nif not os.path.exists(ahf_basename + \'halos\'):\n    raise FileNotFoundError(f"AHF halos file does not exist: {ahf_basename}.AHF_halos")\nelse:\n    print(f"AHF halos file found: {ahf_basename}halos")\nhalos = s.halos(halo_numbers=\'v1\', ahf_basename=ahf_basename)'

## Open Trace

By default, it will save your trace. In my case I've already run this trace, so it will decline to overwrite it. In your case, you'll see information printed to screen as the trace goes on. Since I've already created the trace, I'll go ahead and load it now:

In [4]:
sim_base

'/home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/'

In [ ]:
# trace = pd.read_hdf(sim_base + 'h148.cosmo50PLK.6144g3HbwK1BH.004096/h148.cosmo50PLK.6144g3HbwK1BH.004096.trace_back.hdf5')

trace = pd.read_hdf('/home/selvani/MAP/Sims/cptmarvel.cosmo25cmb/cptmarvel.cosmo25cmb.4096g5HbwK1BH/cptmarvel.cosmo25cmb.4096g5HbwK1BH.004096/cptmarvel.cosmo25cmb.4096g5HbwK1BH.004096.trace_back2.hdf5')

The index is the present day halo ID. Each column represents the main progenitor of a halo at that timestep:

In [5]:
trace

004032  003936  003840  003744  003648  003606  003552  003456  \
004096                                                                    
0040961       1       1       1       1       1       1       1       1   

         003360  003264  ...  000347  000288  000275  000225  000192  000188  \
004096                   ...                                                   
0040961       1       1  ...       1       1       1       1      24      27   

         000139  000107  000096  000071  
004096                                   
0040961      31      -1      -1     NaN  

[1 rows x 61 columns]

If you don't know how to use pandas, this might take a bit to get used to. To see the main progenitor IDs of halo 7, for example, you do this:

In [14]:
trace.loc['0040961']

003968     1
003840     1
003712     1
003636     1
003584     1
003456     1
003328     1
003245     1
003200     1
003072     1
002944     1
002816     1
002688     1
002624     1
002560     1
002432     1
002304     1
002176     1
002162     1
002048     1
001920     1
001813     1
001792     1
001664     1
001543     1
001536     1
001408     1
001331     1
001280     1
001162     1
001152     1
000896     1
000818     1
000768     1
000672     1
000640     1
000512     1
000482     1
000291     2
000199    42
000146    49
000128     3
Name: 0040961, dtype: int64

To see the main progenitor ID just at step 000188, for example, you would do this:

In [6]:
# Fix the index by removing the '004096' prefix
trace.index = trace.index.str.replace('004096', '')

# Convert to integers if desired
trace.index = trace.index.astype(int)

In [7]:
trace

004032  003936  003840  003744  003648  003606  003552  003456  \
004096                                                                   
1            1       1       1       1       1       1       1       1   

        003360  003264  ...  000347  000288  000275  000225  000192  000188  \
004096                  ...                                                   
1            1       1  ...       1       1       1       1      24      27   

        000139  000107  000096  000071  
004096                                  
1           31      -1      -1     NaN  

[1 rows x 61 columns]

In [9]:
trace.to_hdf('/home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/h329.cosmo50PLK.3072gst5HbwK1BH.004096/h329.cosmo50PLK.3072gst5HbwK1BH.004096.trace_back2.hdf5', key='ids')

/tmp/ipykernel_32056/3202716096.py:1: PerformanceWarning: 
your performance may suffer as PyTables will pickle object types that it cannot
map directly to c-types [inferred_type->floating,key->block1_values] [items->Index(['000071'], dtype='object')]

  trace.to_hdf('/home/selvani/MAP/Sims/h329.cosmo50PLK.3072g/h329.cosmo50PLK.3072gst5HbwK1BH/h329.cosmo50PLK.3072gst5HbwK1BH.004096/h329.cosmo50PLK.3072gst5HbwK1BH.004096.trace_back2.hdf5', key='ids')


## Analyze

In [22]:
trace = pd.read_hdf('/home/selvani/MAP/pynbody/cptmarvel.cosmo25cmb.4096g5HbwK1BH.004096.trace_back.hdf5', key='ids')

In [23]:
trace.loc[7, '003840']

np.int64(7)

or for several steps:

In [24]:
trace.loc[7, ['000199', '000640']]

000199    12
000640     8
Name: 7, dtype: int64

The -1s (and, in some cases, NaNs) represent steps where the main progenitor could not be idenified. Most larger halos are traced back pretty far. here is an example of a halo that wasn't:

In [25]:
trace.loc[16]

003968    16
003840    16
003712    16
003636    16
003584    16
003456    16
003328    16
003245    16
003200    17
003072    16
002944    14
002816    14
002688    14
002624    14
002560    14
002432    14
002304    13
002176    14
002162    14
002048    13
001920    13
001813    14
001792    14
001664    13
001543    14
001536    14
001408    14
001331    14
001280    14
001162    19
001152    20
000896    18
000818    16
000768    14
000672    18
000640    19
000512    17
000482    18
000291    19
000199    32
000146    -1
000128    44
Name: 16, dtype: int64

## Do I Need This?

Using the identified halos from the trace, you can read in any value in the amiga.stat or AHF_halos files and put it in a DataFrame of its own. Every time you read in an amiga.stat file you'll get some output information; this mainly has to do with connecting AHF files to amiga files, and you should be able to ignore it.

In [21]:
mvir_trace, rvir_trace, distance = ht.tracing.trace_quantity(sim_base, trace_df=trace, 
                                       quantity=['Mvir(M_sol)', 'Rvir(kpc)', 'Nearest'])

/home/selvani/MAP/Sims/cptmarvel.cosmo25cmb/cptmarvel.cosmo25cmb.4096g5HbwK1BH/cptmarvel.cosmo25cmb.4096g5HbwK1BH.004096/cptmarvel.cosmo25cmb.4096g5HbwK1BH.004096 



/home/selvani/MAP/pynbody/halo_trace/halo_trace/amiga_ahf.py:149: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  df = pd.read_csv(amiga_file,


Rvir is within 0.01%! Match confirmed.


TypeError: loop of ufunc does not support argument 0 of type numpy.float64 which has no callable sqrt method

Again, if you don't know pandas, this might take a bit to get used to. The maximum halo mass reached by every halo throughout its history (at least, its successfully traced history) can be found in one step:

In [10]:
mvir_trace.max(axis=1)

Grp
1        2.402080e+12
2        1.327024e+11
3        6.989641e+10
4        3.851439e+10
5        4.217958e+10
             ...     
15289    2.986070e+06
15291    1.913236e+06
15292    2.574561e+06
15293    2.824864e+06
15294    3.182441e+06
Length: 11356, dtype: float64

In [11]:
mvir_trace.idxmax(axis=1)

Grp
1        004096
2        003360
3        002555
4        002592
5        002880
          ...  
15289    002400
15291    003072
15292    001024
15293    001920
15294    001664
Length: 11356, dtype: object

Or, the mass at z=6 (step 275 in this simulation):

In [12]:
mvir_trace['000275']

Grp
1        2.863878e+08
2        2.216010e+09
3        3.669721e+09
4        7.197241e+09
5        1.117208e+09
             ...     
15289             NaN
15291             NaN
15292             NaN
15293             NaN
15294             NaN
Name: 000275, Length: 11356, dtype: float64

## Troubleshooting

If you're encountering errors related to the AHF or amiga files, it's possible there are multiple versions of the AHF catalogs laying around in the directory. Make sure there is one set of AHF files, and that the amiga files correspond to that run of AHF.